src/services/codefixes/importFixes.ts

```diff
@@ -504,42 +504,60 @@ namespace ts.codefix {
+                        // First remove the extension
+                        let moduleSpecifier = removeFileExtension(moduleFileName);
+                        // If the module could be imported by a directory name, use that directory's name
+                        moduleSpecifier = getDirectoryOrFileName(moduleSpecifier);
+                        // Get a path that's relative to node_modules or the importing file's path
+                        moduleSpecifier = getNodeResolvablePath(moduleSpecifier);
+                        // If the module was found in @types, get the actual node package name
+                        return getPackageNameFromAtTypesDirectory(moduleSpecifier);
+
+                        function getDirectoryOrFileName(fullModulePathWithoutExtension: string): string {
+                            // If the file is the main module, it can be imported by the package name
+                            const indexOfLastNodeModules = moduleFileName.lastIndexOf("node_modules");
+                            const indexOfSlashAtPackageRoot = moduleFileName.indexOf("/", indexOfLastNodeModules + 13 /* "node_modules\".length */);
+                            const packageRootPath = moduleFileName.substring(0, indexOfSlashAtPackageRoot);
+                            const packageJsonPath = combinePaths(packageRootPath, "package.json");
+                            if (context.host.fileExists(packageJsonPath)) {
+                                const packageJsonContent = JSON.parse(context.host.readFile(packageJsonPath));
                                 if (packageJsonContent) {
-                                    const mainFile = packageJsonContent.main || packageJsonContent.typings;
-                                    if (mainFile) {
-                                        const mainExportFile = toPath(mainFile, moduleDirectory, getCanonicalFileName);
+                                    const mainFileRelative = packageJsonContent.typings || packageJsonContent.types || packageJsonContent.main;
+                                    if (mainFileRelative) {
+                                        const mainExportFile = toPath(mainFileRelative, packageRootPath, getCanonicalFileName);
                                         if (removeFileExtension(mainExportFile) === removeFileExtension(moduleFileName)) {
-                                            relativeFileName = getDirectoryPath(relativeFileName);
+                                            return packageRootPath;
                                         }
                                     }
                                 }
                             }
-                            catch (e) { }
+
+                            // If the file is index.js, it can be imported by its directory name
+                            if (endsWith(fullModulePathWithoutExtension, "/index")) {

```

Comment:
I should check on this. Since ```/index``` was existing code, I relied on the assumption that it was normalized -- but wouldn't hurt to verify.